### DISCUSSION TITLE
Customização dos eventos 'OnBeforeExecute' e 'OnAfterExecute'

### DISCUSSION THREAD

**Discussion by diegofgeremias** [viniciussanchez/RESTRequest4Delphi#173] [Ideas]
Gostaria de sugerir uma melhoria:

Possibilitar a parametrização e manipulação dos eventos 'OnBeforeExecute' e 'OnAfterExecute' das requisições executadas.

Seria algo do tipo:

`TRequest.New`
`    .OnBeforeExecute(`
`      procedure (ARequest: IRequest)`
`      begin`
`        ARequest.ContentType('application/json');`
`      end)`
`    .OnAfterExecute(`
`      procedure (ARequest: IRequest; AResponse: IResponse)`
`      begin`
`        if (AResponse.StatusCode <> 200) then`
`          raise Exception.Create('Minha exceção customizada! ' + AResponse.Content);`
`      end)`
`    .BaseURL('http://127.0.0.1:9000/ping')`
`    .Post;`

Esses dois manipuladores possibilitariam a customização da preparação e do recebimento das requisições que são executadas dentro da aplicação. Com a implementação desses dois novos modificadores, seria possível até definir essas duas propriedades de maneira genérica e centralizada dentro da aplicação.

Caso essa funcionalidade seja coerente, providencio o PR.

Desde já obrigado!

**Comment by viniciussanchez** [comment]
Legal, pode enviar o PR sim, será bem vindo. Tente testar no Lazarus também se possível, e usando as diversas engine criadas...
Desde já, obrigado...

**Comment by viniciussanchez** [comment]
O código que você subiu, teve quebra de compatibilidade com o Lazarus. Eu arrumei aqui. Se puder testar novamente suas funcionalidades...

**Reply by diegofgeremias** [reply]
Testei novamente os motores Client, Indy, Synapse e NetHttp com o teu ajuste para compatibilização com Lazarus e aparentemente os eventos estão funcionando como esperado! Muito obrigado!
